test/models/player_test.rb

```diff
@@ -0,0 +1,38 @@
+require 'test_helper'
+
+class PlayerTest < ActiveSupport::TestCase
+  let(:player) do
+    Player.new(
+      game: game,
+      name: "Matt",
+      seat_number: 1,
+      faction: Faction.find_by_key(faction_key),
+    )
+  end
+  let(:faction_key) { "l1z1x" }
+  let(:game) { games(:one) }
+
+  it "creates a valid player" do
+    _(player.valid?).must_equal true
+  end
+
+  describe "Starting Tech" do
+    let(:faction_key) { "arborec" }
+
+    it "copies the correct player.technologies" do
+      player.save!
```

Comment:
Seeing the test code now, I must admit I find it a bit odd to `save` the object to be honest 🤔 What about having starting tech set once the faction is assigned?

Reply:
> Seeing the test code now, I must admit I find it a bit odd to `save` the object to be honest 🤔 What about having starting tech set once the faction is assigned?

Right. I was a bit lazy on that one. I did not want to overwrite `#faction=` and was not sure what the best lifecycle callback would be to set the `starting_tech` but `before_create` is definately sub-optimal.

Reply:
Maybe overriding the `#faction=` is the way to go: 539848f70abd46a57dabb4099853aa96718addf4

How would you do it?

Reply:
IIRC overriding setter in ActiveRecord is no problem (it was in very old Rails versions) and thus it should be fine. Regarding your implementation I would suggest two things:

1. Return `faction` instead of `technologies` (seems a bit odd to get those returned)
2. Maybe it would be more idempotent to union the `Player#technologies` with `Faction#starting_tech`: `self.technologies |= faction.starting_tech`. This should be safe since we default both to empty arrays.

Reply:
> Return faction instead of technologies (seems a bit odd to get those returned)

:+1:

> Maybe it would be more idempotent to union the Player#technologies with Faction#starting_tech: self.technologies |= faction.starting_tech. This should be safe since we default both to empty arrays.
 
I think setting the technologies and therefore resetting existing technologies is semantically closer to what we are modeling.

In a real game a player can only ever pick one faction per Game and cannot have tech before choosing a faction.

I was thinking about having adding some guards to the faction setter because it effectively should be write-once but… yeah, overriding a faction is something that as of now cannot happen through the Web UI/Controllers.

Reply:
You're right. I thought about something else: Loading the player from DB would eventually invoke the setter. But that again is from the jurasic ages of Rails and won't happen anymore (ironically it was the reason why overriding the setter was problematic). Anyhow: Let's wrap this up for now and add guards in a next step.